File: link_list/reorder_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
      stack<ListNode*> store_node;
        ListNode *cur = head;
        int node_count = 0;
        
        if (head == nullptr)
            return;
        
        while (cur != nullptr) {
            node_count++;
            store_node.emplace(cur);
            cur = cur->next;
        }
        cur = head;
        for (int i = 0; i < node_count / 2; i++) {
            ListNode *next = cur->next;
            cur->next = store_node.top();
            store_node.top()->next = next;
            store_node.pop();
            cur = next;
        }
        cur->next = nullptr;
    }
};


//Approach-2
class Solution {
    private:
        ListNode* reverse_link_list(ListNode* start)
        {
            ListNode* current = start;
            ListNode* previous = NULL;
            ListNode* temp;
            while(current)
            {
                temp = current->next;
                current->next = previous;
                previous = current;
                current = temp;
            }
            start = previous;
            return start;
        }
public:
    void reorderList(ListNode* head) {
        if (head == NULL)
            return;
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* first_part = head;
        ListNode* second_part = slow->next;
        slow->next = NULL;
        second_part = reverse_link_list(second_part);
        ListNode* dummy_node = new ListNode(0);
        ListNode* temp = dummy_node;
        while(first_part || second_part)
        {
            if(first_part)
            {
                temp->next = first_part;
                temp = temp->next;
                first_part = first_part->next;
            }
            if(second_part)
            {
                temp->next = second_part;
                temp = temp->next;
                second_part = second_part->next;
            }
            
        }
        dummy_node = dummy_node->next;
        
        
    }
};

File: Meta/Practice/Python/Reorder List.py
'''
ou are given the head of a singly linked-list. The list can be represented as:

L0 → L1 → … → Ln - 1 → Ln
Reorder the list to be on the following form:

L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
You may not modify the values in the list's nodes. Only nodes themselves may be changed.
'''

'''
Constraints :
The number of nodes in the list is in the range [1, 5 * 104].
1 <= Node.val <= 1000
'''

'''
Input: head = [1,2,3,4]
Output: [1,4,2,3]
'''
'''
Complexity :
N = Number of nodes in LL
TC : O(N)
SC : O(1)
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow = head
        fast = head
        while fast and fast.next != None:
            slow = slow.next
            fast = fast.next.next
        
        first = head
        secondhalf = slow.next
        slow.next = None
        secondhalf = self.reverseLinkList(secondhalf)
        dummyNode = ListNode(-1)
        temp = dummyNode
        
        while first or secondhalf:
            if first :
                temp.next = first
                first = first.next
                temp = temp.next
            if secondhalf:
                temp.next = secondhalf
                secondhalf = secondhalf.next
                temp = temp.next

        return dummyNode.next
    
    def reverseLinkList(self,head):
        current = head
        prev = None
        temp = None
        while current != None :
            temp = current.next
            current.next = prev
            prev = current
            current = temp
        head = prev
        return head